src/per/uart.cpp

```diff
@@ -20,11 +24,40 @@ static UartRingBuffer DMA_BUFFER_MEM_SECTION   uart_dma_fifo;
 static void Error_Handler()
 {
     asm("bkpt 255");
+    while(1) {}
 }
 
-// Uses HAL so these things have to be local to this file only
-struct uart_handle
+class UartHandler::Impl
 {
+  public:
+    UartHandler::Result Init(const UartHandler::Config& config);
+
+    const UartHandler::Config& GetConfig() const { return config_; }
+
+    int PollReceive(uint8_t* buff, size_t size, uint32_t timeout);
+
+    UartHandler::Result StartRx();
+
+    bool RxActive();
+
+    UartHandler::Result FlushRx();
+
+    UartHandler::Result PollTx(uint8_t* buff, size_t size);
+
+    uint8_t PopRx();
+
+    size_t Readable();
+
+    int CheckError();
+
+    UartHandler::Result InitPins();
+
+    UartHandler::Result DeInitPins();
+
+    GPIO_TypeDef*  port;
+    USART_TypeDef* periph;
+    uint16_t       tx, rx;
```

Comment:
These shouldn't be necessary.

periph is accessible with huart.Instance without an additional pointer

port implies that only a single port can be used while both pins are below. This could lead to issues if tx/rx are on different ports.

The pins are already located in config_.pins, and I don't think they need to be stored for any particular reason.